ompi/mca/pml/ob1/pml_ob1_sendreq.c

```diff
@@ -262,12 +258,20 @@ mca_pml_ob1_rget_completion (mca_pml_ob1_rdma_frag_t *frag, int64_t rdma_length)
 {
     mca_pml_ob1_send_request_t *sendreq = (mca_pml_ob1_send_request_t *) frag->rdma_req;
     mca_bml_base_btl_t *bml_btl = frag->rdma_bml;
+    size_t frag_remaining;
 
     /* count bytes of user data actually delivered and check for request completion */
     if (OPAL_LIKELY(0 < rdma_length)) {
-        OPAL_THREAD_ADD_FETCH_SIZE_T(&sendreq->req_bytes_delivered, (size_t) rdma_length);
+        frag_remaining = OPAL_THREAD_SUB_FETCH_SIZE_T(&frag->rdma_bytes_remaining, (size_t)rdma_length);
         SPC_USER_OR_MPI(sendreq->req_send.req_base.req_ompi.req_status.MPI_TAG, (ompi_spc_value_t)rdma_length,
                         OMPI_SPC_BYTES_SENT_USER, OMPI_SPC_BYTES_SENT_MPI);
+
+        if( 0 == frag_remaining ) {  /* this frag is now completed. Update the request and be done */
+            OPAL_THREAD_ADD_FETCH_SIZE_T(&sendreq->req_bytes_delivered, frag->rdma_length);
```

Comment:
Just a question.

This protocol may segment a message to fragments and this function (`mca_pml_ob1_rget_completion`) is called per fragment. Before your change, `req_bytes_delivered` was decreased per fragment. After your change, `req_bytes_delivered` is decreased (to zero) when all fragments are transfered. Is it appropriate?

Reply:
Indeed, before each RGET completion was updating the request itself, and never manage the life cycle of the fragment (it was release with the request, so we had a strong limit of one fragment per request). Now the RGET completion works at the fragment level, and decreases it's payload until the fragment is completed and only then it touches the corresponding request. The issue here is that we need to know when the remote got all the data from a fragment in order to be able to clean it. With this change, the fragment cleanup can be done in a single place, and the protocol can now support fragmentation of the RGETs.